usecases/auth/authentication/apikey/dynamic_users.go

```diff
@@ -0,0 +1,161 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2024 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package apikey
+
+import (
+	"crypto/sha256"
+	"crypto/subtle"
+	"fmt"
+	"sync"
+
+	"github.com/alexedwards/argon2id"
+
+	"github.com/weaviate/weaviate/entities/models"
+)
+
+type DynamicUser interface {
+	CreateUser(userId, secureHash, userIdentifier string) error
+	DeleteUser(userId string) error
+	GetUsers(userIds ...string) (map[string]*User, error)
+	RotateKey(userId, secureHash string) error
+	CheckUserIdentifierExists(userIdentifier string) (bool, error)
+}
+
+type User struct {
+	Id                 string
+	Active             bool
+	InternalIdentifier string
+}
+
+type DynamicApiKey struct {
+	sync.RWMutex
+	weakKeyStorageById   map[string][sha256.Size]byte
+	secureKeyStorageById map[string]string
+	identifierToId       map[string]string
+	IdToIdentifier       map[string]string
+	users                map[string]*User
+}
+
+func NewDynamicApiKey() *DynamicApiKey {
+	return &DynamicApiKey{
+		weakKeyStorageById:   make(map[string][sha256.Size]byte),
+		secureKeyStorageById: make(map[string]string),
+		identifierToId:       make(map[string]string),
+		IdToIdentifier:       make(map[string]string),
+		users:                make(map[string]*User),
+	}
+}
+
+func (c *DynamicApiKey) CreateUser(userId, secureHash, userIdentifier string) error {
+	c.Lock()
+	defer c.Unlock()
+	_, secureKeyExists := c.secureKeyStorageById[userId]
+	_, identifierExists := c.identifierToId[userId]
+	_, usersExists := c.users[userId]
+
+	// Todo-Dynuser: Does this make sense? Error would end up with RAFT layer
```

Comment:
The way the 2pc replication logic handles this is with [custom error codes](https://github.com/weaviate/weaviate/blob/main/usecases/replica/transport.go#L43) that are then caught specifically in the coordinator and [re-raised](https://github.com/weaviate/weaviate/blob/main/usecases/replica/replicator.go#L158) appropriately from there.

Can something equivalent be arranged here using [status codes](https://grpc.io/docs/guides/status-codes/#the-full-list-of-status-codes)? @reyreaud-l wdyt?

Reply:
not sure how GRPC status codes are relevant here 😅 

I think what is unclear to me is:
- this code runs locally on a node, after RAFT did its thing
- will the changes be rolled back on other nodes if it does not work on one node? I don't see how that would happen
- This error should never happen (we check for existence before things reach RAFT) but I could see a race happen where the same user is created in different requests. 

I want to put a bit more thought into this, but first ge the basics working

Reply:
Ah, I misunderstood then. I thought this TODO was in relation to throwing errors that need communication between the nodes. If so, status codes could be relevant there to communicate the problem and then throw where necessary

Looking at the RAFT impl, this error would be thrown within the backoff retry logic which would be problematic 🤔 `func (s *schemaHandlers) createTenants` would seem to have the same potential issue as tenant existence(s) are validated before the RAFT request allowing for these types of races

Reply:
maybe we need to add a "reserve" call to fix these races 🤔 

- reserve user_id+identifier via RAFT (with a timeout of the reservation)
- check that it succeeded 
- actually add the user, without any checks